src/System.IO.Pipelines/PipeAwaitable.cs

```diff
@@ -10,16 +10,34 @@ internal struct PipeAwaitable
     {
         private static readonly Action _awaitableIsCompleted = () => { };
         private static readonly Action _awaitableIsNotCompleted = () => { };
+        private static readonly Action<object> _cancelOperation = (state) => { };
 
         private CancelledState _cancelledState;
         private Action _state;
+        private CancellationToken _cancellationToken;
+        private CancellationTokenRegistration _cancellationTokenRegistration;
 
         public PipeAwaitable(bool completed)
         {
             _cancelledState = CancelledState.NotCancelled;
             _state = completed ? _awaitableIsCompleted : _awaitableIsNotCompleted;
         }
 
+
+        public void AttachToken(CancellationToken cancellationToken, Action<object> callback, object state)
+        {
+            if (cancellationToken != _cancellationToken)
+            {
+                _cancellationTokenRegistration.Dispose();
+                _cancellationToken = cancellationToken;
+                if (_cancellationToken.CanBeCanceled)
+                {
+                    _cancellationToken.ThrowIfCancellationRequested();
```

Comment:
This one is so Flush/ReadAsync would throw immediately if passed cancelled token